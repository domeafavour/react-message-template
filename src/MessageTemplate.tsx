import React from "react";
import { TokensRenderer, TokensRendererProps } from "./TokensRenderer";
import { ElementForwardedRef, HTMLElementTypes } from "./typings";
import { useTokens } from "./useTokens";

interface Props<S extends string = string, T extends HTMLElementTypes = "div">
  extends Omit<TokensRendererProps<S>, "tokens"> {
  tag?: T;
  tagProps?: Omit<React.ComponentPropsWithoutRef<T>, "children">;
  template?: S;
}

export const MessageTemplate = React.forwardRef(
  <S extends string = string, T extends HTMLElementTypes = "div">(
    { tag, tagProps, values, template, noMatch }: Props<S, T>,
    forwardedRef: ElementForwardedRef<T>
  ) => {
    const tokens = useTokens<S>(template);
    return React.createElement(
      tag ?? "div",
      { ...tagProps, ref: forwardedRef },
      <TokensRenderer<S> tokens={tokens} values={values} noMatch={noMatch} />
    );
  }
) as <S extends string = string, T extends HTMLElementTypes = "div">(
  props: Props<S, T> & { ref?: ElementForwardedRef<T> }
) => React.ReactNode;
