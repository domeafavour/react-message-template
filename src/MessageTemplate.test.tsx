import { render } from "@testing-library/react";
import React from "react";
import { describe, expect, it, vi } from "vitest";
import { MessageTemplate } from "./MessageTemplate";

describe("MessageTemplate", () => {
  it("should render a string", () => {
    const result = render(<MessageTemplate template="hello world" />);
    expect(result.container).toMatchSnapshot();
  });

  it("should render correctly", () => {
    const result = render(
      <MessageTemplate
        template="hello, {name}"
        values={{ name: <strong>John Doe</strong> }}
      />
    );
    expect(result.container).toMatchSnapshot();
  });

  it("default noMatch(`{key}`)", () => {
    const result = render(
      <MessageTemplate
        template="his name is {name}, and his age is {age}."
        values={{ name: <strong>John Doe</strong> }}
      />
    );
    expect(result.container).toMatchSnapshot();
  });

  it("customized noMatch", () => {
    const result = render(
      <MessageTemplate
        template="his name is {name}, and his age is {age}."
        values={{ name: <strong>John Doe</strong> }}
        noMatch={(key) => (
          <span style={{ color: "red" }}>{`No value for key: ${key}`}</span>
        )}
      />
    );
    expect(result.container).toMatchSnapshot();
  });

  it("render as a code tag", () => {
    const result = render(
      <MessageTemplate
        tag="code"
        template='console.log("{message}")'
        values={{ message: "Hello World" }}
      />
    );
    expect(result.container).toMatchSnapshot();
  });

  it("refCallback should be called", () => {
    const refCallback = vi.fn() as React.RefCallback<HTMLDivElement | null>;
    expect(refCallback).not.toHaveBeenCalled();
    render(<MessageTemplate template="hello world" ref={refCallback} />);
    expect(refCallback).toHaveBeenCalledTimes(1);
  });
});
