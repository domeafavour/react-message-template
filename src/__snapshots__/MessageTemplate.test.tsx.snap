// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageTemplate > customized noMatch 1`] = `
<div>
  <div>
    his name is 
    <strong>
      John Doe
    </strong>
    , and his age is 
    <span
      style="color: red;"
    >
      No value for key: age
    </span>
    .
  </div>
</div>
`;

exports[`MessageTemplate > default noMatch(\`{key}\`) 1`] = `
<div>
  <div>
    his name is 
    <strong>
      John Doe
    </strong>
    , and his age is 
    {age}
    .
  </div>
</div>
`;

exports[`MessageTemplate > render as a code tag 1`] = `
<div>
  <code>
    console.log("
    Hello World
    ")
  </code>
</div>
`;

exports[`MessageTemplate > should render a string 1`] = `
<div>
  <div>
    hello world
  </div>
</div>
`;

exports[`MessageTemplate > should render correctly 1`] = `
<div>
  <div>
    hello, 
    <strong>
      John Doe
    </strong>
  </div>
</div>
`;
